ublox_derive/src/input.rs

```diff
@@ -0,0 +1,643 @@
+                }
+            } else {
+                return Err(syn::Error::new(
+                    list.nested[0].span(),
+                    "Invalid repr attribute for ubx_type enum",
+                ));
+            }
+            syn::parse_quote! { u8 }
+        }
+        _ => {
+            return Err(syn::Error::new(
+                attr.span(),
+                "Invalid repr attribute for ubx_type enum",
+            ))
+        }
+    };
+    let mut variants = Vec::with_capacity(in_variants.len());
+    for var in in_variants {
+        if syn::Fields::Unit != var.fields {
+            return Err(syn::Error::new(
+                var.fields.span(),
+                "Invalid variant for ubx_type enum",
+            ));
+        }
+        let var_sp = var.ident.span();
+        let (_, expr) = var
+            .discriminant
+            .ok_or_else(|| Error::new(var_sp, "ubx_type enum variant should has value"))?;
+        let variant_value = if let syn::Expr::Lit(syn::ExprLit {
+            lit: syn::Lit::Int(litint),
+            ..
+        }) = expr
+        {
+            litint.base10_parse::<u8>()?
+        } else {
+            return Err(syn::Error::new(
+                expr.span(),
+                "Invalid variant value for ubx_type enum",
+            ));
+        };
+        variants.push((var.ident, variant_value));
+    }
+
+    let attrs = attrs
+        .into_iter()
+        .filter(|x| !x.path.is_ident("ubx") && !x.path.is_ident("ubx_extend"))
+        .collect();
+
+    Ok(UbxExtendEnum {
+        attrs,
+        name: enum_name,
+        repr,
+        from_fn,
+        into_fn,
+        rest_handling,
+        variants,
+    })
+}
+
+pub fn parse_bitflags(mac: syn::ItemMacro) -> syn::Result<BitFlagsMacro> {
+    let (from_fn, into_fn, rest_handling) =
+        parse_ubx_extend_attrs("#[ubx_extend_bitflags]", mac.span(), &mac.attrs)?;
+
+    let ast: BitFlagsAst = syn::parse2(mac.mac.tokens)?;
+
+    let valid_types: [(Type, u32); 3] = [
+        (syn::parse_quote!(u8), 1),
+        (syn::parse_quote!(u16), 2),
+        (syn::parse_quote!(u32), 4),
+    ];
+    let nbits = if let Some((_ty, size)) = valid_types.iter().find(|x| x.0 == ast.repr_ty) {
+        size * 8
+    } else {
+        let mut valid_type_names = String::with_capacity(200);
+        for (t, _) in &valid_types {
+            if !valid_type_names.is_empty() {
+                valid_type_names.push_str(", ");
+            }
+            valid_type_names.push_str(&t.into_token_stream().to_string());
+        }
+        return Err(Error::new(
+            ast.repr_ty.span(),
+            format!("Not supported type, expect one of {:?}", valid_type_names),
+        ));
+    };
+
+    let mut consts = Vec::with_capacity(ast.items.len());
+    for item in ast.items {
+        consts.push(BitFlagsMacroItem {
+            attrs: item.attrs,
+            name: item.name,
+            value: item.value.base10_parse()?,
+        });
+    }
```

Comment:
Can you do this using `ast.items.iter().map(...).collect()`?

Reply:
> ast.items.iter().map(...).collect()?

It is hard, because of `value: item.value.base10_parse()?`, see **?** at the end.

Reply:
I haven't tried it, but I think collect() should be able to collect over Result values, stopping on the first error: https://doc.rust-lang.org/rust-by-example/error/iter_result.html

Reply:
> I haven't tried it, but I think collect() should be able to collect over Result values

1. This variant would be slower: rust-lang/rust#48994
2. It require more lines, and more complex code:
```rust
    let consts = ast
        .items
        .into_iter()
        .map(|item| {
            Ok(BitFlagsMacroItem {
                attrs: item.attrs,
                name: item.name,
                value: item.value.base10_parse()?,
            })
        })
        .collect::<syn::Result<Vec<_>>>()?;
```

11 lines vs 8, plus slower. Is it worth this?

Reply:
I don't think performance really matters here, but I'm fine with dropping.

Reply:
@lkolbly 

> I don't think performance really matters here

The funny thing that performance really matter in build script and procedural macroses,
specially performance in debug mode.

Because of:
1. Debug vs Release difference in Rust case is really huge, may be 20x, may be 100x difference
2. By default procedural macroses / build scripts compiled with the same optimization
    level as the main crate.

So if some day packet.rs will contain all packages from ubx spec, compilation of ublox
in debug mode may become slow. But good thing is that at least nightly cargo allow change optimization level for specific dependencies, so this is not as bad as before.

Reply:
Even if packet.rs has every packet, it's still on the order of thousands of items at most - pre-allocating vs. not-pre-allocating the vector holding them will be an unnoticeable performance difference to the end user.

Reply:
My comment was in general about performance of whole ublox_derive, not about specific `with_capacity`. If talk about this specific case **not** inlined iterators methods (because of debug) vs simple loop may create difference for thousands of packets.